// ytdlp-nodejs/lib/EmitterEvent.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
class EmitterEvent {
    constructor() {
        this.listeners = {};
    }
    addListener(event, listener) {
        this.listeners[event] = this.listeners[event] || [];
        this.listeners[event].push(listener);
        return this;
    }
    on(event, listerner) {
        return this.addListener(event, listerner);
    }
    removeListerner(event, listerner) {
        let lis = this.listeners[event];
        if (!lis)
            return this;
        for (let i = lis.length; i > 0; i--) {
            const element = lis[i];
            if (element === listerner) {
                lis.splice(i, 1);
                break;
            }
        }
        return this;
    }
    off(event, listener) {
        return this.removeListerner(event, listener);
    }
    once(event, listener) {
        this.listeners = this.listeners[event] || [];
        const onceWrapper = () => {
            listener();
            this.off(event, onceWrapper);
        };
        this.listeners[event].push(onceWrapper);
        return this;
    }
    emit(event, ...args) {
        let fns = this.listeners[event];
        if (event === "error" && !fns) {
            throw new Error(args[0]);
        }
        if (!fns)
            return false;
        fns.forEach((f) => {
            f(...args);
        });
        return true;
    }
    listernerCount(event) {
        let fns = this.listeners[event] || [];
        return fns.length;
    }
    rawListerner(event) {
        return this.listeners[event];
    }
}
exports.default = EmitterEvent;
